File: JAVA/EstudosP2/ExercicioLista01.java
/**
 * EXERCÍCIO 1 - Remover todos os elementos pares de uma lista simples flexível
 * Complexidade: Theta(n) — percorre a lista uma única vez.
 *
 * LÓGICA:
 * - Dois ponteiros: 'ant' (anterior) e 'atual'.
 * - Se atual.elemento é par: ant.prox = atual.prox (pula o nó).
 * - Se não é par: apenas avança ant.
 * - A célula-cabeça garante que nunca tentamos remover antes do primeiro nó real.
 */
class Celula1 {
    int elemento;
    Celula1 prox;
    public Celula1() { prox = null; }
    public Celula1(int e) { elemento = e; prox = null; }
}

class Lista1 {
    Celula1 primeiro, ultimo;

    public Lista1() {
        primeiro = ultimo = new Celula1(); /* celula-cabeca */
    }

    public void inserirFim(int x) {
        ultimo.prox = new Celula1(x);
        ultimo = ultimo.prox;
    }

    public void mostrar() {
        System.out.print("[ ");
        for (Celula1 i = primeiro.prox; i != null; i = i.prox) {
            System.out.print(i.elemento + " ");
        }
        System.out.println("]");
    }

    /* ===== IMPLEMENTAÇÃO ===== */
    public void removerPares() {
        Celula1 ant   = primeiro;          /* começa na cabeça */
        Celula1 atual = primeiro.prox;

        while (atual != null) {
            if (atual.elemento % 2 == 0) {
                /* elemento é par: remover */
                ant.prox = atual.prox;

                /* se removemos o último, atualizar ponteiro */
                if (atual == ultimo) {
                    ultimo = ant;
                }
                /* ant não avança; atual avança */
                atual = ant.prox;
            } else {
                /* elemento é ímpar: apenas avança */
                ant   = atual;
                atual = atual.prox;
            }
        }
    }
}

public class ExercicioLista01 {
    public static void main(String[] args) {
        Lista1 l = new Lista1();
        l.inserirFim(1);
        l.inserirFim(2);
        l.inserirFim(3);
        l.inserirFim(4);
        l.inserirFim(5);
        l.inserirFim(6);

        System.out.print("Antes:  ");
        l.mostrar();

        l.removerPares();

        System.out.print("Depois: ");
        l.mostrar();
        /* esperado: [ 1 3 5 ] */
    }
}

File: JAVA/EstudosP2/exercicio_lista01.c
/**
 * EXERCÍCIO 1 - Remover todos os elementos pares de uma lista simples flexível
 * Complexidade: Theta(n) — percorre a lista uma única vez.
 * 
 * LÓGICA:
 * - Dois ponteiros: 'ant' (anterior) e 'atual'.
 * - Se atual.elemento é par: ant.prox = atual.prox (pula o nó), libera o nó.
 * - Se não é par: apenas avança ant.
 * - A célula-cabeça garante que nunca tentamos remover antes do primeiro nó real.
 */

#include <stdio.h>
#include <stdlib.h>

/* ===== ESTRUTURA ===== */
typedef struct Celula {
    int elemento;
    struct Celula *prox;
} Celula;

typedef struct {
    Celula *primeiro; /* celula-cabeca */
    Celula *ultimo;
} Lista;

/* ===== FUNÇÕES AUXILIARES ===== */
void start(Lista *l) {
    l->primeiro = (Celula*) malloc(sizeof(Celula));
    l->primeiro->prox = NULL;
    l->ultimo = l->primeiro;
}

void inserirFim(Lista *l, int x) {
    Celula *nova = (Celula*) malloc(sizeof(Celula));
    nova->elemento = x;
    nova->prox = NULL;
    l->ultimo->prox = nova;
    l->ultimo = nova;
}

void mostrar(Lista *l) {
    printf("[ ");
    Celula *i;
    for (i = l->primeiro->prox; i != NULL; i = i->prox) {
        printf("%d ", i->elemento);
    }
    printf("]\n");
}

/* ===== IMPLEMENTAÇÃO ===== */
void removerPares(Lista *l) {
    Celula *ant = l->primeiro;   /* começa na cabeça */
    Celula *atual = l->primeiro->prox;

    while (atual != NULL) {
        if (atual->elemento % 2 == 0) {
            /* elemento é par: remover */
            ant->prox = atual->prox;

            /* se removemos o último, atualizar ponteiro */
            if (atual == l->ultimo) {
                l->ultimo = ant;
            }

            free(atual);
            atual = ant->prox; /* avança sem mover 'ant' */
        } else {
            /* elemento é ímpar: apenas avança */
            ant = atual;
            atual = atual->prox;
        }
    }
}

/* ===== TESTE ===== */
int main(void) {
    Lista l;
    start(&l);

    inserirFim(&l, 1);
    inserirFim(&l, 2);
    inserirFim(&l, 3);
    inserirFim(&l, 4);
    inserirFim(&l, 5);
    inserirFim(&l, 6);

    printf("Antes: ");
    mostrar(&l);

    removerPares(&l);

    printf("Depois: ");
    mostrar(&l);

    /* esperado: [ 1 3 5 ] */
    return 0;
}
